sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp

```diff
@@ -0,0 +1,373 @@
+//===------- matrix-unified.hpp - SYCL matrix extension ----*- C++ -*------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+
+#pragma once
+#include <sycl/ext/oneapi/matrix/matrix-tensorcores.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+namespace matrix {
+
+// forward declarations
+template <typename T, use Use, size_t Rows, size_t Cols,
+          layout Layout = layout::dynamic, typename Group = sycl::sub_group>
+struct joint_matrix;
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols, use Use,
+          layout Layout, typename T2>
+inline __SYCL_ALWAYS_INLINE void
+joint_matrix_fill(Group sg,
+                  joint_matrix<T, Use, NumRows, NumCols, Layout, Group> &res,
+                  const T2 &v);
+
+template <
+    typename Group, typename S, typename T, use Use, size_t NumRows,
+    size_t NumCols, matrix::layout Layout, access::address_space Space,
+    access::decorated IsDecorated,
+    std::enable_if_t<std::is_same<S, std::remove_const_t<T>>::value ||
+                         (std::is_same<S, precision::tf32>::value &&
+                          std::is_same<std::remove_const_t<T>, float>::value),
+                     bool> = true>
+inline __SYCL_ALWAYS_INLINE void
+joint_matrix_load(Group sg,
+                  joint_matrix<S, Use, NumRows, NumCols, Layout, Group> &res,
+                  multi_ptr<T, Space, IsDecorated> src, size_t stride);
+
+template <
+    typename Group, typename S, typename T, size_t NumRows, size_t NumCols,
+    access::address_space Space, access::decorated IsDecorated,
+    std::enable_if_t<std::is_same<S, std::remove_const_t<T>>::value, bool> =
+        true>
+inline __SYCL_ALWAYS_INLINE void joint_matrix_load(
+    Group sg,
+    joint_matrix<S, use::accumulator, NumRows, NumCols,
+                 sycl::ext::oneapi::experimental::matrix::layout::dynamic,
+                 Group> &res,
+    multi_ptr<T, Space, IsDecorated> src, size_t stride,
+    sycl::ext::oneapi::experimental::matrix::layout Layout);
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          access::address_space Space, access::decorated IsDecorated>
+inline __SYCL_ALWAYS_INLINE void joint_matrix_store(
+    Group sg,
+    joint_matrix<T, use::accumulator, NumRows, NumCols,
```

Comment:
so, we don't provide matrix_store for A&B in unified matrix?

Reply:
Yes that's correct: at least not in the main "[sycl_ext_oneapi_matrix](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc)" extension. The main extension is for features that can be used across backends. Store for A&B is not even possible to implement as a software solution in Nvidia backends because the mapping from registers to the matrix array to be stored isn't known. Also Store for A&B doesn't appear to have a common use case (or at least one anyone here is aware of yet).

It could still be added in the intel specific matrix extension if desired.

BTW I've updated this PR with your suggestion from https://github.com/intel/llvm/pull/7203#issuecomment-1299690682
I've moved all `joint_matrix_cuda` implementations to the cuda implementation header. I've also made the `joint_matrix_cuda` member (analogous to your `__spv::__spirv_JointMatrixINTEL`) in `joint_matrix` private because I don't think this should be available to the user directly. Due to this I have made `joint_matrix_load` `joint_matrix_store` `joint_matrix_mad` `joint_matrix_fill` friends of `joint_matrix`.

I've kept the different implementation (wrt Intel backends) of the `wi_data` class in the cuda backend but made use of return type inference to keep a single `get_wi_data()`:

```
#if defined(__SYCL_DEVICE_ONLY__)
  inline __SYCL_ALWAYS_INLINE decltype(auto) get_wi_data() {
#if defined(__NVPTX__)
    return wi_data(cuda_impl.wi_marray);
#else
//  intel impl: return wi_data<T, NumRows, NumCols, Use, Layout, Group>(*this);
#endif
  };
#endif
```

So now I think that everything is fine for the portability across Intel/nvptx.

What do you think?

Thanks

Reply:
> so, we don't provide matrix_store for A&B in unified matrix?

These Intel-specific features should be kept in a separate file (matrix-jit). Similar to what Jack did: we can move matrix-jit.hpp to matrix-jit-legacy.hpp and leave the Intel specific features (use based) under matrix-jit.hpp.
matrix-jit-use.hpp should be deleted as everything there will be achieved from matrix-unified and matrix-jit

Reply:
Ideally, we don't want to include such target specific that's why we are introducing such separation.
We will know more once we are done using matrix-unified.hpp for the JIT implementation (hopefully by end of this week).
@yubingex007-a11y, do you know whether we will run in such issues on our side?

Reply:
LGTM. it is much more elegant than previous version.
BTW, why get_wi_data() is wrapped by "#if defined(__SYCL_DEVICE_ONLY__)"?

> Yes that's correct: at least not in the main "[sycl_ext_oneapi_matrix](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc)" extension. The main extension is for features that can be used across backends. Store for A&B is not even possible to implement as a software solution in Nvidia backends because the mapping from registers to the matrix array to be stored isn't known. Also Store for A&B doesn't appear to have a common use case (or at least one anyone here is aware of yet).
> 
> It could still be added in the intel specific matrix extension if desired.
> 
> BTW I've updated this PR with your suggestion from [#7203 (comment)](https://github.com/intel/llvm/pull/7203#issuecomment-1299690682) I've moved all `joint_matrix_cuda` implementations to the cuda implementation header. I've also made the `joint_matrix_cuda` member (analogous to your `__spv::__spirv_JointMatrixINTEL`) in `joint_matrix` private because I don't think this should be available to the user directly. Due to this I have made `joint_matrix_load` `joint_matrix_store` `joint_matrix_mad` `joint_matrix_fill` friends of `joint_matrix`.
> 
> I've kept the different overload (wrt Intel backends) of the `wi_data` class in the cuda backend but made use of return type inference to keep a single `get_wi_data()`:
> 
> ```
> #if defined(__SYCL_DEVICE_ONLY__)
>   inline __SYCL_ALWAYS_INLINE decltype(auto) get_wi_data() {
> #if defined(__NVPTX__)
>     return wi_data(cuda_impl.wi_marray);
> #else
> //  intel impl: return wi_data<T, NumRows, NumCols, Use, Layout, Group>(*this);
> #endif
>   };
> #endif
> ```
> 
> So now I think that everything is fine for the portability across Intel/nvptx.
> 
> What do you think?
> 
> Thanks